unified-runtime/source/adapters/opencl/event.hpp

```diff
@@ -0,0 +1,84 @@
+//===--------- queue.hpp - OpenCL Adapter ---------------------------===//
+//
+// Copyright (C) 2023 Intel Corporation
+//
+// Part of the Unified-Runtime Project, under the Apache License v2.0 with LLVM
+// Exceptions. See LICENSE.TXT
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+#pragma once
+
+#include "common.hpp"
+#include "queue.hpp"
+
+#include <vector>
+
+struct ur_event_handle_t_ {
+  using native_type = cl_event;
+  native_type CLEvent;
+  ur_context_handle_t Context;
+  ur_queue_handle_t Queue;
+  std::atomic<uint32_t> RefCount = 0;
+  bool IsNativeHandleOwned = true;
+
+  ur_event_handle_t_(native_type Event, ur_context_handle_t Ctx,
+                     ur_queue_handle_t Queue)
+      : CLEvent(Event), Context(Ctx), Queue(Queue) {
+    RefCount = 1;
+    urContextRetain(Context);
+    if (Queue) {
+      urQueueRetain(Queue);
+    }
+  }
+
+  ~ur_event_handle_t_() {
+    urContextRelease(Context);
+    if (Queue) {
+      urQueueRelease(Queue);
+    }
+    if (IsNativeHandleOwned) {
+      clReleaseEvent(CLEvent);
+    }
+  }
+
+  uint32_t incrementReferenceCount() noexcept { return ++RefCount; }
+
+  uint32_t decrementReferenceCount() noexcept { return --RefCount; }
+
+  uint32_t getReferenceCount() const noexcept { return RefCount; }
+
+  ur_result_t ensureQueue() {
+    if (!Queue) {
+      cl_command_queue native_queue;
+      CL_RETURN_ON_FAILURE(clGetEventInfo(CLEvent, CL_EVENT_COMMAND_QUEUE,
+                                          sizeof(native_queue), &native_queue,
+                                          nullptr));
+      UR_RETURN_ON_FAILURE(ur_queue_handle_t_::makeWithNative(
+          native_queue, Context, nullptr, Queue));
+    }
+
+    return UR_RESULT_SUCCESS;
+  }
+};
+
+inline cl_event *ifUrEvent(ur_event_handle_t *returnedEvent, cl_event &event) {
+  return returnedEvent ? &event : nullptr;
+}
+
+inline ur_result_t createUREvent(cl_event event, ur_context_handle_t context,
+                                 ur_queue_handle_t queue,
+                                 ur_event_handle_t *returnedEvent) {
```

Comment:
Nitpick that we're mixing variable naming conventions in this file, e.g. `ur_queue_handle_t queue` here but `ur_queue_handle_t Queue` in the `ur_event_handle_t_` constructor.